### ISSUE TITLE
Orb color for UtBS's disengage ability (unit can move but can't attack)

### ISSUE THREAD

**Issue by stevecotton** [wesnoth/wesnoth#5155]
UtBS and some UMC abilities allow a unit to attack and then move, although it can't attack again. Currently the unit gets the partially-moved orb color, the same as a unit that can still attack. I'm wondering about adding a new orb color specifically for this state, shown in this prototype image with white and teal:
![wesnoth_disengage_white_teal](https://user-images.githubusercontent.com/101462/93460858-2ed27d80-f8e4-11ea-938c-8bc3982ffb19.png)

The orb colors themselves are configurable in advanced preferences, the main change here would be to add the extra state to the orb-choosing code. Adding documentation for this would overlap with #4455.

Now that I have an issue number, a commit with some some very early experimental code will get pushed and auto-linked to this issue by Github.

@Pentarctagon: I'd be interested in putting this on the 1.15.7 roadmap.
@nemaara: as UtBS is the main driver for this, how do you feel about it?

**Comment by cooljeanius** [comment]
This always confused me when playing thru UtBS and a new orb color would definitely help; thanks for filing this!

**Comment by AI0867** [comment]
I'm not opposed to this, but do consider the case where `max_attacks` is larger than 1. In that case, you may end up with `0<attacks_left<max_attacks`. I would say that as the unit can still attack, it should have a yellow orb, and only once `attacks_left` has hit 0, should the orb use the new color, but other opinions could be defensible.

Somewhat related question: Is it `max_attacks` or `attacks` or both? [UnitTypeWML](https://wiki.wesnoth.org/UnitTypeWML) says `attacks`, [SingleUnitWML](https://wiki.wesnoth.org/SingleUnitWML) says `max_attacks` and I thought the keys were supposed to match these days.

**Comment by Pentarctagon** [comment]
@stevecotton given UtBS is the main reason for this, I'll add it to the roadmap if nemaara is good with the idea.

**Comment by CelticMinstrel** [comment]
So we have an orb colour when a unit can attack and move, and an orb colour when a unit cannot attack or move. This new one is for when a unit can move but not attack. Is there one for when a unit can attack but not move? I think that currently uses the partial colour, right? Should we add another new colour for that state?

**Comment by stevecotton** [comment]
> consider the case where max_attacks is larger than 1. In that case, you may end up with 0<attacks_left<max_attacks. I would say that as the unit can still attack, it should have a yellow orb, and only once attacks_left has hit 0, should the orb use the new color

Agreed, and I think my implementation already does it this way (although I've yet to test that).

> Should we add another new colour for when a unit can attack but not move?

I'm not opposed to that, although I'm worried that we'll end up with too many colors. I think the situation isn't as important as the disengage orb, because it's something that's already visible on the map - a unit with a partial orb adjacent to an enemy (with the edge case that it's a non-skirmisher unit). With the disengage orb, it's typical to move the unit out of the way so that another unit can attack, and so you can have a map with no visual clue of whether the unit has already attacked.

**Comment by doofus-01** [comment]
> I'm not opposed to that, although I'm worried that we'll end up with too many colors.

I also wonder if more colors is really the best method.  Is it difficult to add another changing element, like a hard outline for the orb if unit can still attack?
![orb-example](https://user-images.githubusercontent.com/13510196/93723500-d2759500-fb53-11ea-919e-4e59c1560603.png)
 (Mock-up just to show what I mean)

**Comment by stevecotton** [comment]
Using a different image for the orb would certainly be possible, and simple with the existing code. From a code perspective, the orb is an image that's the size of a whole hex and which already uses Image Path Functions for the recoloring, so nothing restricts new artwork to being an orb. I'm refactoring the code to reduce the amount of copy-pasted code for each color, but will ensure this is still simple with the new code.

If we do allow different images, then it would need a redesign of the associated preferences ("Preferences", "Advanced", "Customize orb colors"). I've no idea how many users customize that, however I've already assumed that it will be quicker to do it than to get through code review without doing so (footnote). Redesigning it is tempting anyway, as it currently has a separate boolean for whether to show each type of orb.

The minimap has a mode that shows orb colors ("Toggle Minimap Unit Coding", not to be confused with "Toggle Minimap Unit Drawing"); the selection of (colors and images) needs to still make sense without the images.

(footnote) This isn't a complaint about those code reviews - learning from the feedback is much more valuable to me a rubber-stamp review would be.

**Comment by stevecotton** [comment]
Having multiple colors would also be possible - maybe one each for (can move) and (can attack).

**Comment by stevecotton** [comment]
Current status of this - I was about to open the PR, and in rereading these comments realised that I'd forgotten @doofus-01's suggestion of using different icons instead of colors. Going to think it over a little more.

**Comment by stevecotton** [comment]
![wesnoth_disengage_boots](https://user-images.githubusercontent.com/101462/107066788-aeeb7800-67de-11eb-9522-e0d0e9250426.png)

@doofus-01 I'd push it back to you for better artwork, but here's a prototype of using a pair of boots instead of an orb.
 
This is in the `0c3238c` commit that's auto-linked above this comment.  In that the color is configurable but not the shape, and I'd prefer to remove the configurability of the colors as opposed to adding addition preferences for the shape; I feel there are already too many options in the preferences related to orbs.

The prototype is using a magenta-recolored rescale of `images/cursors/move.png`.

**Comment by doofus-01** [comment]
I see the logic behind the boots, but I wonder if there isn't reason to keep the orb shape, at least to some extent?  Changing to a small item/icon makes it look like another decoration, like a crown, indicating something new.  It also might be hard to read at a glance, especially if there are more non-orb movement-icons (and if there aren't, it will probably look a bit odd).  My suggestion up above wasn't all that intuitive either.  Would making a half-orb recolor be an option?  I agree that more, customizable colors isn't the answer.
![Screenshot_20210206_130155](https://user-images.githubusercontent.com/13510196/107129860-6636cc00-687d-11eb-9f22-ef7fdf637d5b.png)
(I didn't just slice orb in half because that might make it look like "half moves left" or similar.)

That's my take in any case, but I don't have UX expertise, and maybe I'm being too conservative.  If this can be cleanly reverted, maybe we should just test it out, see what the feedback is?

**Comment by stevecotton** [comment]
![wesnoth_disengage_two_color](https://user-images.githubusercontent.com/101462/107157085-71275480-6982-11eb-9e53-c860d66a97c3.png)

I like that idea. I made it use image-recoloring so that the existing preferences for orb colors would also affect it. 

As an accident of the image editing, the red is bright like the partially-moved orb instead of the dull red of the full-moved orb; I think that's a bonus for making it different to the red orb. However, I'd welcome a better image if someone makes one.

**Comment by CelticMinstrel** [comment]
Okay, I have two comments on that latest commit.

* I think the two-colour orb is wrong. The upper half is hard-coded to red, but I think it should use the configured colour for the unmoved orb. This probably requires splitting it into two separate images, so it might be a bit complicated… perhaps a new issue? I don't know if it's a good idea, but this would also allow flipping the logic - if you can attack but can't move, colour the top half yellow and the bottom half red.
* The help text is a bit confusing about when the yellow vs two-colour orb is used. After reading it a few times, I think two-colours is "can't attack but can move" while yellow is "can attack and maybe move", so the key difference is whether you can attack, right?

**Comment by doofus-01** [comment]
The colors are both customizable, nothing is hard-coded to red.

**Comment by CelticMinstrel** [comment]
Hmm, so I'm guessing the top colour uses a different source palette? I guess it would be ellipse_red. That does make sense.

The comment about maybe flipping the orb still stands.

**Comment by stevecotton** [comment]
> The comment about maybe flipping the orb still stands.

I've done a prototype for an "engaged" orb, it's autolinked above.

I'm not completely opposed to the idea, but feel the engaged state is something that's already visible on the map without needing an extra orb - a unit with a partial orb that's adjacent to an enemy is in the engaged state (with the edge case that it's a non-skirmisher unit). With the disengage state, it's typical to move the unit out of the way so that another unit can attack, and so without the disengage orb there's no visual clue of whether the unit has already attacked.

**Comment by stevecotton** [comment]
Replying to @AI0867's comment in the already-merged PR, https://github.com/wesnoth/wesnoth/issues/5523#issuecomment-779401647

> Going by the code, that would result in the following table?
> 
> ```
> moves\attacks | 0   | some   | all
> --------------+-----+--------+-----
> 0             | red | R/O    | R/O
> some          | O/R | orange | orange
> all           | O/R | orange | green
> ```

Yes, the prototyped code in 617c2bc would give the table above. The first row isn't exactly zero movement points, it will also be shown for a unit with 1 MP but with every adjacent hex's terrain requiring 2 MP. There should be a check for the teleport ability too, but it wasn't there before and it seemed low enough priority that I just added a todo comment.

> I guess we can interpret this is orange/red being independent, but green being the "unit fully fresh" state.
> 
> One could consider (though it would be a much bigger change) to fully decouple the two, giving us the following grid:
> 
> ```
> moves\attacks | 0   | some   | all
> --------------+-----+--------+-----
> 0             | red | R/O    | R/G
> some          | O/R | orange | O/G
> all           | G/R | G/O    | green
> ```

My question to all of this is whether additional color combinations are giving additional useful information to the player. For both the engage orb and the half-green orbs, how will the additional combination help the player?

I'm opposed to the half-green orbs, because when I'm playing I read the green orb as "unit will rest heal". That isn't what the code checks, but it seems reliable - the only edge cases that I can think of involve BCoA's or SXC's movement potions.

**Comment by CelticMinstrel** [comment]
So with the first system, the "engage" orb means you can attack but not move, "disengage" means you can move but not attack, and orange means you can do both. It would be useless to see "engage" if you're not already next to an enemy, since you can't move next to one in order to attack, so for the sake of this example let's assume that you _are_ next to an enemy.

* If you have the "engage" orb, your only option is to attack
* If you have an orange orb, you can attack, or you can choose to move away from the enemy
* If you have the "disengage" orb, you can't attack, but you can move away

I'm not sure _how_ useful this is. Certainly, if the "engage" orb was just pure orange, you'd be able to tell that the unit can move away when you click and its movement range is highlighted, so the information is already there, but not instantly accessible.

---

Half-green orbs don't seem as useful as the half-red ones. The distinction between "full" and "partial" just doesn't matter very much. In fact, you could get rid of green orbs entirely and I don't think much would be lost (other than what you mentioned about rest healing).

**Comment by AI0867** [comment]
Okay, given all that, I'm in favor of going with 617c2bc.

**Comment by CelticMinstrel** [comment]
I'm also not sure of the term "engaged orb" but I also can't think of anything better, so I guess we'll have to stick with it…

**Comment by stevecotton** [comment]
> Okay, given all that, I'm in favor of going with 617c2bc.

Why, as in, "what about the engaged orb improves the UX enough to be worth having"? Do you want it enough to own the PR for it, and to fix up the artwork (of the two red/orange color combinations, I feel that only one of them looks good)?

I want the disengaged orb, but the engaged orb feels more like "we could do this" rather than "this improves the UX". I'm also worried that the engaged orb looks too similar to the disengaged orb, so it creates visual noise that reduces the usefulness of the disengaged orb.

**Comment by AI0867** [comment]
I'd have to play with it for a bit, but I'm guessing it would help with determining whether a unit can still move to attack elsewhere, or just still attack. Due to ZoC this isn't often relevant, but if there are level 0 units, skirmishers or similar abilities around, they can prevent confusion there.

**Comment by doofus-01** [comment]
> I want the disengaged orb, but the engaged orb feels more like "we could do this" rather than "this improves the UX". I'm also worried that the engaged orb looks too similar to the disengaged orb, so it creates visual noise that reduces the usefulness of the disengaged orb.

I agree, making too many combos ruins the idea.

**Comment by stevecotton** [comment]
> Commit 617c2bc does not appear to be on any branch. Is there a way I can check it out without manually applying the patch?

It's on the `engaged_orb` branch in my fork. Can be grabbed with:
```
git fetch https://github.com/stevecotton/wesnoth engaged_orb
git cherry-pick FETCH_HEAD
```

Note - if you try to merge it instead of cherry-picking, you'll get a merge conflict because that branch has an older version of the HDPI preferences changes.
